File: 卡码/代码随想录/1/07-回溯/Go/06-其他/LeetCode_0491.go
//491. 非递减子序列
package main

import "slices"

func findSubsequences(nums []int) [][]int {
	res:=make([][]int,0)
	path:=make([]int,0)
	var backtracking func(nums []int,index int)
	backtracking=func(nums []int,index int){
		if len(path)>1 {
			tmp:=make([]int,len(path))
			copy(tmp,path)
			res=append(res,tmp)
		}
		visited:=make([]bool,201)
		for i:=0;i<len(visited);i++ {
			visited[i]=false
		}
		for i:=index;i<len(nums);i++ {
			if len(path)>0&&nums[i]<path[len(path)-1]||visited[nums[i]+100] {
				continue
			}
			visited[nums[i]+100]=true
			path=append(path,nums[i])
			backtracking(nums,i+1)
			path=path[:len(path)-1]
		}
	}
	backtracking(nums,0)
	return res
}

File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/回溯/_06_其他/LeetCode_0491.java
package com.kurumi.codeThoughts.回溯._06_其他;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 491. 非递减子序列
 */
public class LeetCode_0491 {
    static class Solution {
        List<List<Integer>> res = new ArrayList<>();
        Deque<Integer> path = new LinkedList<>();

        public void backtracking(int[] nums, int index) {
            if (path.size() > 1)
                res.add(new ArrayList(path));
            boolean[] visited = new boolean[201];
            Arrays.fill(visited, false);
            for (int i = index; i < nums.length; i++) {
                if (!path.isEmpty() && nums[i] < path.getLast() || visited[nums[i] + 100])
                    continue;
                visited[nums[i] + 100] = true;
                path.offer(nums[i]);
                backtracking(nums, i + 1);
                path.removeLast();
            }
        }

        public List<List<Integer>> findSubsequences(int[] nums) {
            backtracking(nums, 0);
            return res;
        }
    }
}
